### PR TITLE
Add the new uninitialized buffer String initializer privately and use it to fix a perf TODO in append

### PR DESCRIPTION
**Pull request by Catfish-Man**
Fixes https://bugs.swift.org/browse/SR-10556

### CODE REVIEW

**Comment by Catfish-Man** [comment]
This subsumes all of https://github.com/apple/swift/pull/23409 *except* for making the new initializer public, and adopting it in Foundation/moving String bridging code into Foundation.

**Comment by Catfish-Man** [comment]
Also this is still failing tests from the append changes. In particular this invariant is being tripped

    _internalInvariant(totalCount > _SmallString.capacity)

which implies that somehow we're failing to make a small string despite the contents fitting, as far as I can tell.

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
That was the wrong PR >.<

**Comment by palimondo** [comment]
The addition of new benchmarks should land in a separate PR, so that you can measure the impact of the perf optimizations done here. Otherwise there will be no *before* baseline to compare against.

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
Oops, still had some merge issues apparently

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
Dangit, nope, that's not going to work. The test run that was looking good showed failures just after I kicked off swift-ci :(

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
Not entirely sure what swift-ci is doing heh. I'll just wait and see if some complete builds appear

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
yay there we go

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by shahmishal** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
That looks much better :)

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by milseman** [comment]
I thought the initializer was approved through SE and landed. Or is this about using it for grow operations?

File: stdlib/public/core/StringGutsRangeReplaceable.swift [hunk 1]
```diff
@@ -97,18 +97,19 @@ extension _StringGuts {
...
   private mutating func _foreignGrow(_ n: Int) {
-    // TODO(String performance): Skip intermediary array, transcode directly
-    // into a StringStorage space.
-    let selfUTF8 = Array(String(self).utf8)
-    selfUTF8.withUnsafeBufferPointer {
-      self = _StringGuts(__StringStorage.create(
-        initializingFrom: $0, capacity: n, isASCII: self.isASCII))
+    let newString = String(uninitializedCapacity: n) { buffer in
+      guard let count = _foreignCopyUTF8(into: buffer) else {
+       fatalError("String capacity was smaller than required")
```
Comment (milseman) [inline stdlib/public/core/StringGutsRangeReplaceable.swift]: 2 space indentation

File: stdlib/public/core/StringGutsRangeReplaceable.swift [hunk 2]
```diff
@@ -127,11 +128,13 @@ extension _StringGuts {
...
       sufficientCapacity = false
     }
+        
     if self.isUniqueNative && sufficientCapacity {
       return
     }
-
-    let totalCount = self.utf8Count + otherCount
+    
+    //If we have to resize anyway, and we fit in smol, we should have made one
```
Comment (milseman) [inline stdlib/public/core/StringGutsRangeReplaceable.swift]: Space after //

File: stdlib/public/core/StringGutsRangeReplaceable.swift [hunk 3]
```diff
@@ -155,22 +158,51 @@ extension _StringGuts {
...
 
-    if self.isSmall && slicedOther._guts.isSmall {
+    let otherCount = slicedOther.utf8Count
+    let totalCount = utf8Count + otherCount
+    let hasEnoughUsableSpace = isUniqueNative &&
+      nativeUnusedCapacity! >= otherCount
+    let shouldBeSmol = totalCount <= _SmallString.capacity &&
+      (isSmall || !hasEnoughUsableSpace)
+    
+    //Avoid throwing away a previous reserveCapacity(), but try to be smol
```
Comment (milseman) [inline stdlib/public/core/StringGutsRangeReplaceable.swift]: Space after //

File: stdlib/public/core/StringGutsRangeReplaceable.swift [hunk 4]
```diff
@@ -155,22 +158,51 @@ extension _StringGuts {
...
   internal mutating func append(_ slicedOther: _StringGutsSlice) {
     defer { self._invariantCheck() }
 
-    if self.isSmall && slicedOther._guts.isSmall {
+    let otherCount = slicedOther.utf8Count
+    let totalCount = utf8Count + otherCount
+    let hasEnoughUsableSpace = isUniqueNative &&
+      nativeUnusedCapacity! >= otherCount
+    let shouldBeSmol = totalCount <= _SmallString.capacity &&
+      (isSmall || !hasEnoughUsableSpace)
```
Comment (milseman) [inline stdlib/public/core/StringGutsRangeReplaceable.swift]: Could you explain the logic here, ideally in a comment. There's a fair amount of and, ors, and nots.

**Comment by milseman** [comment]
@swift-ci please benchmark

**Review by milseman — APPROVED** [review]
LGTM, with some nitpick formatting and a request for comments or minor logic cleanup.

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
Oh that's interesting… those benchmark regressions sure weren't there before. Will take a look.

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
Well THAT's fun. I fixed the regression and got a 2-3x speedup instead, but… more work to be done.

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringGutsRangeReplaceable.swift [hunk 5]
```diff
@@ -147,30 +150,84 @@ extension _StringGuts {
 
   internal mutating func append(_ other: _StringGuts) {
     if self.isSmall && other.isSmall {
-      if let smol = _SmallString(self.asSmall, appending: other.asSmall) {
+      if let smol = _SmallString(self.asSmall, appending:
```
Comment (Catfish-Man) [inline stdlib/public/core/StringGutsRangeReplaceable.swift]: whoops messed up line break here while fixing perf

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Pull request merged**